tensorflow_encrypted/tensor/odd_implicit.py

```diff
@@ -0,0 +1,137 @@
+        return self.add(other)
+
+    def __sub__(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        return self.sub(other)
+
+    def __mul__(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        return self.mul(other)
+
+    def __mod__(self, k: int) -> 'OddImplicitTensor':
+        return self.mod(k)
+
+    def add(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        x, y = _lift(self, self.dtype), _lift(other, self.dtype)
+
+        def func(vals):
+            x, y = vals
+
+            negative = lambda: tf.cond(x < tf.int32.min - y, lambda: x + y - 1, lambda: x + y)
+
+            maybe_negative = lambda: tf.cond(y < 0, negative, lambda: x + y)
+
+            positive = lambda: tf.cond(x > tf.int32.max - y, lambda: x + y + 1, lambda: x + y)
+
+            return tf.cond(y > 0, positive, maybe_negative)
+
+        ret = tf.map_fn(func, (x.value, y.value), dtype=self.dtype)
+
+        return OddImplicitTensor(ret, dtype=self.dtype)
+
+    def sub(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        x, y = _lift(self, self.dtype), _lift(other, self.dtype)
+
+        def func(vals):
+            x, y = vals
+
+            positive = lambda: tf.cond(x < tf.int32.min + y, lambda: x - y - 1, lambda: x - y)
+
+            maybe_positive = lambda: tf.cond(y < 0, positive, lambda: x - y)
+
+            negative = lambda: tf.cond(x > tf.int32.max + y, lambda: x - y + 1, lambda: x - y)
+
+            return tf.cond(y < 0, negative, maybe_positive)
+
+        ret = tf.map_fn(func, (x.value, y.value), dtype=self.dtype)
+
+        return OddImplicitTensor(ret, dtype=self.dtype)
+
+    def mul(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        raise NotImplementedError()
+
+    def dot(self, other: Union['OddImplicitTensor', int]) -> 'OddImplicitTensor':
+        raise NotImplementedError()
+
+    def im2col(self, h_filter, w_filter, padding, strides) -> 'OddImplicitTensor':
+        raise NotImplementedError()
+
+    def conv2d(self, other, strides, padding='SAME') -> 'OddImplicitTensor':
+        raise NotImplementedError()
+
+    def mod(self, k: int) -> 'OddImplicitTensor':
+        x = _lift(self, self.dtype)
+        return OddImplicitTensor(x.value % k, self.dtype)
+
+    def transpose(self, perm: Union[List[int], Tuple[int]]) -> 'OddImplicitTensor':
+        return OddImplicitTensor(tf.transpose(self.value, perm), self.dtype)
+
+    def strided_slice(self, args: Any, kwargs: Any) -> 'OddImplicitTensor':
+        return OddImplicitTensor(tf.strided_slice(self.value, *args, **kwargs), self.dtype)
+
+    def reshape(self, axes: Union[tf.Tensor, List[int]]) -> 'OddImplicitTensor':
+        return OddImplicitTensor(tf.reshape(self.value, axes), self.dtype)
+
+
+def _lift(x: Union['OddImplicitTensor', int], dtype) -> 'OddImplicitTensor':
```

Comment:
you could alternatively move this guy into the class and use the `self` parameter instead of passing `dtype`